File: Packages/TimeControlCore/Sources/TimeControlCore/WeekMath.swift
import Foundation

/// Week numbering for a term. Week 1 starts on the Monday of the week containing `termStart`.
public struct TermWeeks: Hashable, Sendable {
    public let termStart: DayKey
    public let termEnd: DayKey

    public init(termStart: DayKey, termEnd: DayKey) {
        self.termStart = termStart
        self.termEnd = max(termEnd, termStart)
    }

    /// Monday of week 1.
    public var firstWeekStart: DayKey { termStart.weekStart }

    /// Number of (possibly partial) weeks the term spans.
    public var weekCount: Int { (termEnd - firstWeekStart) / 7 + 1 }

    /// 1-based week number of `day`, or nil when `day` is outside `termStart...termEnd`.
    public func weekNumber(of day: DayKey) -> Int? {
        guard day >= termStart, day <= termEnd else { return nil }
        return (day - firstWeekStart) / 7 + 1
    }

    /// Monday of week `n` (1-based). Not clamped to the term.
    public func weekStart(ofWeek n: Int) -> DayKey {
        firstWeekStart + 7 * (n - 1)
    }

    /// The seven days of week `n`, not clamped to the term.
    public func days(inWeek n: Int) -> ClosedRange<DayKey> {
        let start = weekStart(ofWeek: n)
        return start...(start + 6)
    }

    /// The days of weeks `first...last`, clamped to the term. Used for "clear weeks 8–9".
    public func days(inWeeks range: ClosedRange<Int>) -> ClosedRange<DayKey> {
        let lo = max(weekStart(ofWeek: range.lowerBound), termStart)
        let hi = min(weekStart(ofWeek: range.upperBound) + 6, termEnd)
        return lo...max(lo, hi)
    }
}

public enum WeekMath {
    /// The instant at `minute` minutes past local midnight on `day`. DST-safe (never `startOfDay + minutes`).
    public static func instant(day: DayKey, minute: Int, calendar: Calendar = .app) -> Date {
        let c = day.civil
        let comps = DateComponents(year: c.year, month: c.month, day: c.day, hour: minute / 60, minute: minute % 60)
        return calendar.date(from: comps)!
    }

    /// Minutes since local midnight of `date`'s day.
    public static func minuteOfDay(_ date: Date, calendar: Calendar = .app) -> Int {
        let c = calendar.dateComponents([.hour, .minute], from: date)
        return c.hour! * 60 + c.minute!
    }

    /// "10:30" style label for a minute-of-day, using the calendar's locale conventions.
    public static func timeLabel(minute: Int, calendar: Calendar = .app) -> String {
        let date = instant(day: .epoch, minute: minute, calendar: calendar)
        return date.formatted(.dateTime.hour().minute())
    }
}

File: web/src/core/weekMath.ts
/**
 * Week numbering for a term, and the DST-safe day/minute to instant conversion.
 * Ported from `WeekMath.swift`.
 */

import { civil, weekStart, type DayKey } from './dayKey'

/** Week 1 starts on the Monday of the week containing `termStart`. */
export interface TermWeeks {
  readonly termStart: DayKey
  readonly termEnd: DayKey
}

export function termWeeks(termStart: DayKey, termEnd: DayKey): TermWeeks {
  return { termStart, termEnd: Math.max(termEnd, termStart) }
}

/** Monday of week 1. */
export function firstWeekStart(t: TermWeeks): DayKey {
  return weekStart(t.termStart)
}

/** Number of (possibly partial) weeks the term spans. */
export function weekCount(t: TermWeeks): number {
  return Math.trunc((t.termEnd - firstWeekStart(t)) / 7) + 1
}

/** 1-based week number of `day`, or null when `day` falls outside the term. */
export function weekNumber(t: TermWeeks, day: DayKey): number | null {
  if (day < t.termStart || day > t.termEnd) return null
  return Math.trunc((day - firstWeekStart(t)) / 7) + 1
}

/** Monday of week `n` (1-based). Not clamped to the term. */
export function weekStartOfWeek(t: TermWeeks, n: number): DayKey {
  return firstWeekStart(t) + 7 * (n - 1)
}

/** The seven days of week `n`, not clamped to the term. */
export function daysInWeek(t: TermWeeks, n: number): [DayKey, DayKey] {
  const start = weekStartOfWeek(t, n)
  return [start, start + 6]
}

/** The days of weeks `first` through `last`, clamped to the term. Backs "clear weeks 8-9". */
export function daysInWeeks(t: TermWeeks, first: number, last: number): [DayKey, DayKey] {
  const lo = Math.max(weekStartOfWeek(t, first), t.termStart)
  const hi = Math.min(weekStartOfWeek(t, last) + 6, t.termEnd)
  return [lo, Math.max(lo, hi)]
}

/**
 * The instant at `minute` minutes past local midnight on `day`.
 *
 * DST-safe: it builds the local wall-clock time rather than adding seconds to midnight, so 10:00 on
 * a spring-forward day is 10:00 and not 11:00.
 */
export function instant(day: DayKey, minute: number): Date {
  const c = civil(day)
  return new Date(c.year, c.month - 1, c.day, Math.trunc(minute / 60), minute % 60, 0, 0)
}

/** Minutes since local midnight of `date`'s day. */
export function minuteOfDay(date: Date): number {
  return date.getHours() * 60 + date.getMinutes()
}
